Mathlib/SetTheory/Ordinal/Arithmetic.lean

```diff
@@ -335,6 +338,60 @@ theorem limitRecOn_limit {C} (o H₁ H₂ H₃ h) :
   simp_rw [limitRecOn, SuccOrder.limitRecOn_limit _ _ h.isSuccLimit, dif_neg h.1]
 #align ordinal.limit_rec_on_limit Ordinal.limitRecOn_limit
 
+/-- Bounded recursion on ordinals. Similar to `limitRecOn`, with the assumption `o < l`
+  added to all cases. The final term's domain is the ordinals below `l`. -/
+@[elab_as_elim]
+def boundedLimitRec {l : Ordinal} (hLim : l.IsLimit) {C : Π o < l, Sort*} (H₁ : C 0 hLim.pos)
+    (H₂ : ∀ o, (h : o < l) → C o h → C (o + 1) (hLim.succ_lt h))
+    (H₃ : ∀ o, (h : o < l) → IsLimit o → (∀ o', (h' : o' < o) → C o' (h'.trans h)) → C o h) :
+    (o : Ordinal) → (h : o < l) → C o h :=
+  fun o ↦ limitRecOn o (fun _ ↦ H₁) (fun o ih h ↦ H₂ _ _ <| ih <| (lt_succ o).trans h)
+    fun _o ho ih _ ↦ H₃ _ _ ho fun _o' h ↦ ih _ h _
+
+@[simp]
+theorem boundedLimitRec_zero {l} (lLim : l.IsLimit) {C} (H₁ H₂ H₃) :
+    @boundedLimitRec l lLim C H₁ H₂ H₃ 0 lLim.pos = H₁ := by
+  simp_all only [boundedLimitRec, limitRecOn_zero, id_eq,
+    eq_mpr_eq_cast, eq_mp_eq_cast, cast_cast, cast_eq]
+
+@[simp]
+theorem boundedLimitRec_succ {l} (lLim : l.IsLimit) {C} (o ho H₁ H₂ H₃) :
+    @boundedLimitRec l lLim C H₁ H₂ H₃ (succ o) (lLim.2 o ho) = H₂ o ho
+    (@boundedLimitRec l lLim C H₁ H₂ H₃ o ho) := by
+  simp_all only [add_one_eq_succ, boundedLimitRec, eq_mpr_eq_cast, eq_mp_eq_cast,
+    limitRecOn_succ, dite_true, cast_cast, cast_eq]
+
+@[simp]
+theorem boundedLimitRec_limit {l : Ordinal} (lLim : l.IsLimit) {C} (o oltl H₁ H₂ H₃ oLim) :
+    @boundedLimitRec l lLim C H₁ H₂ H₃ o oltl = H₃ o oltl oLim (fun x hx ↦
+    @boundedLimitRec l lLim C H₁ H₂ H₃ x (hx.trans oltl)) := by
+  simp_all only [boundedLimitRec, eq_mpr_eq_cast, eq_mp_eq_cast, limitRecOn_limit,
+    dite_true, cast_cast, cast_eq]
+
+/-- Bounded recursion on the ordinals with a constant return type. Similar to `boundedLimitRec`,
+  but with a constant motive function. Useful for defining function to ordinals. -/
+def boundedLimitRec' {α : Sort*} {l : Ordinal} (lLim : l.IsLimit) (H₁ : α)
```

Comment:
Is this really useful? 20 extra lines of API to save on a few characters doesn't seem like a good tradeoff to me.